# Data Scientist Salaries - Data Processing and analysis  

In [323]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn as skl
from sklearn.preprocessing import LabelEncoder

## Data Preparation

In [324]:
df = pd.read_csv('ds_salaries.csv')
print(df.shape)
print('\n')
df.info()

(3755, 11)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


In [325]:
df.head()

work_year experience_level employment_type                 job_title   
0       2023               SE              FT  Principal Data Scientist  \
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio   
0   80000             EUR          85847                 ES           100  \
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [326]:
df.dtypes

work_year              int64
experience_level      object
employment_type       object
job_title             object
salary                 int64
salary_currency       object
salary_in_usd          int64
employee_residence    object
remote_ratio           int64
company_location      object
company_size          object
dtype: object

In [327]:
# Extract categorical variables
cat_vars = ['experience_level', 'employment_type', 'job_title', 'employee_residence', 'company_location', 'company_size']

# Create LabelEncoder object
encoder = LabelEncoder()

# Encode categorical variables as integers
for var in cat_vars:
    df[var] = encoder.fit_transform(df[var])

In [328]:
df.head()

work_year  experience_level  employment_type  job_title  salary   
0       2023                 3                2         84   80000  \
1       2023                 2                0         66   30000   
2       2023                 2                0         66   25500   
3       2023                 3                2         47  175000   
4       2023                 3                2         47  120000   

  salary_currency  salary_in_usd  employee_residence  remote_ratio   
0             EUR          85847                  26           100  \
1             USD          30000                  75           100   
2             USD          25500                  75           100   
3             USD         175000                  11           100   
4             USD         120000                  11           100   

   company_location  company_size  
0                25             0  
1                70             2  
2                70             2  
3                12             1  
4                12             1

In [329]:
df.drop(['salary','salary_currency'] , axis='columns', inplace = True)
df.head()

work_year  experience_level  employment_type  job_title  salary_in_usd   
0       2023                 3                2         84          85847  \
1       2023                 2                0         66          30000   
2       2023                 2                0         66          25500   
3       2023                 3                2         47         175000   
4       2023                 3                2         47         120000   

   employee_residence  remote_ratio  company_location  company_size  
0                  26           100                25             0  
1                  75           100                70             2  
2                  75           100                70             2  
3                  11           100                12             1  
4                  11           100                12             1

## Data Transformation

In [330]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df_scaled = pd.DataFrame(scaler.fit_transform(df), columns = df.columns)

## Feature Selection

## Regression - Salary Prediction

In this section, a regression model is going to be trained and tested to predict the salaries of Data Scientists based on the other featuress.

In [331]:
#df_scaled converted to dataframe
df_scaled = pd.DataFrame(df_scaled, columns = df.columns)

# Describing Data Set
df_scaled.describe()

work_year  experience_level  employment_type     job_title   
count  3.755000e+03      3.755000e+03     3.755000e+03  3.755000e+03  \
mean  -1.593130e-13      1.967948e-16     4.475189e-16 -5.109096e-17   
std    1.000133e+00      1.000133e+00     1.000133e+00  1.000133e+00   
min   -3.433303e+00     -2.725008e+00    -1.495172e+01 -2.139921e+00   
25%   -5.404380e-01     -5.178456e-01     2.592668e-02 -6.831569e-01   
50%   -5.404380e-01      5.857357e-01     2.592668e-02 -3.594315e-01   
75%    9.059945e-01      5.857357e-01     2.592668e-02  3.959278e-01   
max    9.059945e-01      5.857357e-01     7.514748e+00  2.823868e+00   

       salary_in_usd  employee_residence  remote_ratio  company_location   
count   3.755000e+03        3.755000e+03  3.755000e+03      3.755000e+03  \
mean    6.055224e-17        2.119329e-16  1.513806e-17     -9.082837e-17   
std     1.000133e+00        1.000133e+00  1.000133e+00      1.000133e+00   
min    -2.100622e+00       -3.439432e+00 -9.524327e-01     -3.550953e+00   
25%    -6.752143e-01        4.601861e-01 -9.524327e-01      4.506246e-01   
50%    -4.076928e-02        4.601861e-01 -9.524327e-01      4.506246e-01   
75%     5.936757e-01        4.601861e-01  1.105918e+00      4.506246e-01   
max     4.955485e+00        5.641760e-01  1.105918e+00      5.077900e-01   

       company_size  
count  3.755000e+03  
mean  -6.812127e-17  
std    1.000133e+00  
min   -2.343022e+00  
25%    2.078761e-01  
50%    2.078761e-01  
75%    2.078761e-01  
max    2.758774e+00

As can be seen, the data is standardly scaled and is ready for regression.

In the following cell, data is split into train and test sets using `sklearn.train_test_split`.

In [332]:
#import train_test_split
from sklearn.model_selection import train_test_split

#seperating features and target
features = df.drop('salary_in_usd', axis=1)
target = df['salary_in_usd'].values

#split dataset into training and test features and targets
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.3, random_state=42, shuffle=False)

Scaling the training and test sets in order to start training. The test sets are scaled using the paramaters gained from fitting the scaler to the training sets. This decreases bias in the testing of the model.

In [333]:
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

#reshaping y_train & y_test to apply the scaling transformation
y_train = y_train.reshape(-1,1)
y_test = y_test.reshape(-1,1)

y_train = scaler.fit_transform(y_train)
y_test = scaler.transform(y_test)


In the following cells, the model is built and tested.

In [334]:
#importing required library to build a neural network model
import tensorflow as tf
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense

In [335]:
#Exploring X_train
print(X_train.shape)
df2 = pd.DataFrame(X_train)
df2.head()

(2628, 8)


0         1          2         3         4         5         6   
0  0.677128  0.539190   0.018394  2.399939 -2.538503  1.328351 -2.549286  \
1  0.677128 -0.626099 -24.151786  1.431901  0.376549  1.328351  0.373349   
2  0.677128 -0.626099 -24.151786  1.431901  0.376549  1.328351  0.373349   
3  0.677128  0.539190   0.018394  0.410082 -3.430866  1.328351 -3.393603   
4  0.677128  0.539190   0.018394  0.410082 -3.430866  1.328351 -3.393603   

          7  
0 -3.531991  
1  3.840584  
2  3.840584  
3  0.154297  
4  0.154297

In [336]:
#Exploring y_train
df3 = pd.DataFrame(y_test)
df3.head()

0
0  0.679168
1 -0.245661
2  1.318053
3  0.197146
4  0.909559

A `Sequential` model is built comprising four `Dense` (fully interconnected) layers. It uses the non-linear `RELU` activation function for the hidden layers, and a `linear` output layer. `L2 kernel regularization` is introduced in order to prevent overfitting and promote good generalization to test scenarios.

In [340]:
#constructing model
model = Sequential([
    tf.keras.Input(shape=(8,)),
    Dense(units = 25, activation = 'relu', kernel_regularizer=tf.keras.regularizers.L2(0.001)),
    Dense(units = 10, activation = 'relu', kernel_regularizer=tf.keras.regularizers.L2(0.001)),
    Dense(units = 5, activation = 'relu', kernel_regularizer=tf.keras.regularizers.L2(0.001)),
    Dense(units = 1, activation = 'linear')
])

Model will use the `Adam`(adaptive moment estimation) optimization algorithm, and the `Mean Squared Error` loss function. It is initialized with a `Learning Rate` of 0.0001.

In [344]:
#compiling and summarizing model
from keras.optimizers import Adam
from keras.losses import MeanSquaredError
model.compile(optimizer = Adam(learning_rate = 0.0001), loss='mean_squared_error')
model.summary()

Model: "sequential_20"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_80 (Dense)            (None, 25)                225       
                                                                 
 dense_81 (Dense)            (None, 10)                260       
                                                                 
 dense_82 (Dense)            (None, 5)                 55        
                                                                 
 dense_83 (Dense)            (None, 1)                 6         
                                                                 
Total params: 546
Trainable params: 546
Non-trainable params: 0
_________________________________________________________________


Now the model is ready to be fit to the data. It is run for a 1000 `epochs`.

In [345]:
history = model.fit(X_train, y_train, validation_split= 0.2 ,epochs = 1000)

Epoch 1/1000
66/66 [==============================] - 1s 4ms/step - loss: 0.6782 - val_loss: 0.7266
Epoch 2/1000
66/66 [==============================] - 0s 2ms/step - loss: 0.6725 - val_loss: 0.7232
Epoch 3/1000
66/66 [==============================] - 0s 2ms/step - loss: 0.6682 - val_loss: 0.7204
Epoch 4/1000
66/66 [==============================] - 0s 2ms/step - loss: 0.6645 - val_loss: 0.7174
Epoch 5/1000
66/66 [==============================] - 0s 2ms/step - loss: 0.6614 - val_loss: 0.7160
Epoch 6/1000
66/66 [==============================] - 0s 2ms/step - loss: 0.6591 - val_loss: 0.7141
Epoch 7/1000
66/66 [==============================] - 0s 2ms/step - loss: 0.6566 - val_loss: 0.7122
Epoch 8/1000
66/66 [==============================] - 0s 2ms/step - loss: 0.6542 - val_loss: 0.7117
Epoch 9/1000
66/66 [==============================] - 0s 2ms/step - loss: 0.6524 - val_loss: 0.7102
Epoch 10/1000
66/66 [==============================] - 0s 2ms/step - loss: 0.6511 - val_loss: 0.7088

Training finished with loss of = 0.6279 on the training set. It can be said that the model has `low bias`, i.e. fits the training data well. However, the validation loss suggests some over fittin, for it has a 12.5% increase over the training loss. The model will be further tested for generalization and overfitting over the test set.

Forward propogation will be used to predict the target for the test set `y_hat`, which will be compared with the true target `y_test`

In [346]:
y_hat = model.predict(X_test)

36/36 [==============================] - 0s 1ms/step


The `mean_squared_error` function is imported from `sklearn` to be applied between `y_hat` and `y_test`

In [347]:
#
from sklearn.metrics import mean_squared_error
mean_squared_error(y_pred = y_hat, y_true = y_test)

1.046112949801076

As can be seen, test error is 1.0461, which compared to the training error of 0.6279 is a 67% increase in error. The model can be said to have high varianve, further improvement is needed.

Real values of `y_hat` and `y_test` can be compared by applying the inverse transformation to the data.

In [348]:
y_hat_real = pd.DataFrame(scaler.inverse_transform(y_hat))
y_test_real = pd.DataFrame(scaler.inverse_transform(y_test))
pd.concat([y_hat_real, y_test_real], axis = 1)

0         0
0     128263.101562  185900.0
1     128263.101562  129300.0
2     140999.406250  225000.0
3     140999.406250  156400.0
4     176624.781250  200000.0
...             ...       ...
1122  147385.078125  412000.0
1123   55704.539062  151000.0
1124   32315.505859  105000.0
1125  -35826.773438  100000.0
1126   31858.498047   94665.0

[1127 rows x 2 columns]

It can be seen the model has a high error margin. There are few things that can be improved. 
1. The model architecture can be changed, such as the number of hidden layers and/or neurons.
2. The hyperparametres can be adjusted, such the learning rate or regularization parametres.
3. More training epochs might be needed.
4. Further data collection and preperation might be needed.